File: service/app/src/test/java/io/miragon/blueprint/adapter/inbound/rest/GlobalExceptionConfigurationTest.java
package io.miragon.blueprint.adapter.inbound.rest;

import io.miragon.blueprint.application.port.inbound.ReportHandoverUseCase;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.operaton.bpm.engine.MismatchingMessageCorrelationException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.willThrow;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;

@WebMvcTest(ReportHandoverController.class)
public class GlobalExceptionConfigurationTest {

    @Autowired
    private MockMvc mockMvc;

    @MockitoBean
    private ReportHandoverUseCase useCase;

    @Test
    @DisplayName("MismatchingMessageCorrelationException maps to 409 with problem+json")
    public void mismatchingMessageCorrelationExceptionMapsTo409WithProblemJson() throws Exception {

        // given: a correlation that fails because the token is no longer at the expected wait state
        String pathVar = "123e4567-e89b-12d3-a456-426614174000";
        willThrow(new MismatchingMessageCorrelationException("miravelo.handoverReported", "No matching token"))
            .given(useCase).reportHandover(any());

        // when: the POST is performed
        MvcResult response =
            mockMvc
                .perform(post("/api/bike-leasing/{applicationId}/report-handover", pathVar))
                .andReturn();

        // then: 409 Conflict is returned as application/problem+json
        assertThat(response.getResponse().getStatus()).isEqualTo(409);
        assertThat(response.getResponse().getContentType()).contains(MediaType.APPLICATION_PROBLEM_JSON_VALUE);
    }
}

File: service/app/src/test/kotlin/io/miragon/blueprint/adapter/inbound/rest/GlobalExceptionConfigurationTest.kt
package io.miragon.blueprint.adapter.inbound.rest

import com.ninjasquad.springmockk.MockkBean
import io.miragon.blueprint.application.port.inbound.ReportHandoverUseCase
import io.mockk.every
import org.assertj.core.api.Assertions.assertThat
import org.operaton.bpm.engine.MismatchingMessageCorrelationException
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post

@WebMvcTest(ReportHandoverController::class)
class GlobalExceptionConfigurationTest {

    @Autowired
    private lateinit var mockMvc: MockMvc

    @MockkBean
    private lateinit var useCase: ReportHandoverUseCase

    @Test
    fun `MismatchingMessageCorrelationException maps to 409 with problem+json`() {

        // given: a correlation that fails because the token is no longer at the expected wait state
        val pathVar = "123e4567-e89b-12d3-a456-426614174000"
        every { useCase.reportHandover(any()) } throws
            MismatchingMessageCorrelationException("miravelo.handoverReported", "No matching token")

        // when: the POST is performed
        val response =
            mockMvc
                .perform(post("/api/bike-leasing/{applicationId}/report-handover", pathVar))
                .andReturn()

        // then: 409 Conflict is returned as application/problem+json
        assertThat(response.response.status).isEqualTo(409)
        assertThat(response.response.contentType).contains(MediaType.APPLICATION_PROBLEM_JSON_VALUE)
    }
}
